### PR TITLE
Add tests for db upgrades

### PR DESCRIPTION
**Pull request by msm-code**
Also fix few small nitpicks about db storage and upgrade format,
and remove max_mem_usage from config (it was never used anyway).

### CODE REVIEW

File: teste2e/test_format_upgrades.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                    "byte_offset": 0,
+                    "file_offset": 0,
+                    "total_files": 0,
+                },
+                name=itermeta,
+            )
+
+    ursadb = subprocess.Popen([ursadb, db, backend])
+    # TODO remove this after we have some sort of quit command
+    time.sleep(1)
```
Comment (icedevml) [inline teste2e/test_format_upgrades.py:70]: :thinking:

Comment (icedevml) [reply teste2e/test_format_upgrades.py:70]: this doesn't really guarantee that the operation succeeded, maybe:
1. launch database
2. `ursacli -q -c status`
3. when the above finished - terminate the DB

ursacli is backed by zmq which will gracefully wait till the database starts to be operational (at least as far as I remember)

File: libursa/Database.cpp [hunk 2]
```diff
@@ -99,19 +93,22 @@ void Database::save() {
...
 
-    json iterators_json;
+    json iterators_json = std::map<std::string, std::string>();
     for (const auto &it : iterators) {
         iterators_json[it.first] = it.second.get_name().get_filename();
     }
     db_json["iterators"] = iterators_json;
+    // Explicit version instead of ursadb_version, because this is is a
+    // *file format version*, not necessarily equal to the db version.
+    db_json["version"] = "1.3.2";
```
Comment (icedevml) [inline libursa/Database.cpp:111]: maybe as a constant and together with other similar constants?

File: teste2e/test_format_upgrades.py (hunk 1 above)
Comment (msm-code) [reply teste2e/test_format_upgrades.py:70]: :thinking: right, completely forgot that I can just ping the DB.

File: libursa/Database.cpp (hunk 2 above)
Comment (msm-code) [reply libursa/Database.cpp:111]: :+1:

File: libursa/Database.cpp [hunk 3]
```diff
@@ -99,19 +94,20 @@ void Database::save() {
     db_file.open(db_base / tmp_db_name, std::ofstream::binary);
 
     json db_json;
-    db_json["config"] = {{"max_mem_size", max_memory_size}};
+    db_json["config"] = std::map<std::string, std::string>();
```
Comment (BonusPlay) [inline libursa/Database.cpp:97]:
```suggestion
    db_json["config"] = std::unordered_map<std::string, std::string>();
```

This requires to `#include <unordered_map>`!

File: libursa/Database.cpp [hunk 4]
```diff
@@ -99,19 +94,20 @@ void Database::save() {
...
 
     std::vector<std::string> dataset_names;
     for (const auto *ds : working_datasets) {
         dataset_names.push_back(ds->get_name());
     }
     db_json["datasets"] = dataset_names;
 
-    json iterators_json;
+    json iterators_json = std::map<std::string, std::string>();
```
Comment (BonusPlay) [inline libursa/Database.cpp:105]:
```suggestion
    json iterators_json = std::unordered_map<std::string, std::string>();
```

This requires to `#include <unordered_map>`!

File: libursa/DatabaseSnapshot.cpp [hunk 5]
```diff
@@ -12,14 +12,12 @@ DatabaseSnapshot::DatabaseSnapshot(
     fs::path db_name, fs::path db_base,
     std::map<std::string, OnDiskIterator> iterators,
     std::vector<const OnDiskDataset *> datasets,
-    const std::map<uint64_t, std::unique_ptr<Task>> &tasks,
-    size_t max_memory_size)
+    const std::map<uint64_t, std::unique_ptr<Task>> &tasks)
```
Comment (BonusPlay) [inline libursa/DatabaseSnapshot.cpp:15]:
```suggestion
    const std::unordered_map<uint64_t, std::unique_ptr<Task>> &tasks)
```

This requires to `#include <unordered_map>`!

File: libursa/DatabaseSnapshot.h [hunk 6]
```diff
@@ -40,8 +39,7 @@ class DatabaseSnapshot {
     DatabaseSnapshot(fs::path db_name, fs::path db_base,
                      std::map<std::string, OnDiskIterator> iterators,
                      std::vector<const OnDiskDataset *> datasets,
-                     const std::map<uint64_t, std::unique_ptr<Task>> &tasks,
-                     size_t max_memory_size);
+                     const std::map<uint64_t, std::unique_ptr<Task>> &tasks);
```
Comment (BonusPlay) [inline libursa/DatabaseSnapshot.h:42]:
```suggestion
                     const std::unordered_map<uint64_t, std::unique_ptr<Task>> &tasks);
```

File: libursa/DatabaseUpgrader.cpp [hunk 7]
```diff
@@ -38,11 +38,27 @@ void save_json(std::string_view path, const json &dbjson) {
...
+    if (db.find("config") != db.end()) {
+        db["config"] = std::map<std::string, std::string>();
+    }
+
+    if (db["config"].find("max_mem_size") != db["config"].end()) {
+        db["config"].erase("max_mem_size");
+    }
+
+    if (db["iterators"] == nullptr) {
+        db["iterators"] = std::map<std::string, std::string>();
```
Comment (BonusPlay) [inline libursa/DatabaseUpgrader.cpp:54]:
```suggestion
        db["iterators"] = std::unordered_map<std::string, std::string>();
```


This requires to `#include <unordered_map>`!

File: teste2e/test_format_upgrades.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+    def tmpfile(self, name=None) -> Path:
+        if name is None:
+            filepath = self.tmpdir + "/" + os.urandom(8).hex()
+        else:
+            filepath = self.tmpdir + "/" + name
+        self.tmpfiles.append(filepath)
+        return Path(filepath)
+
+    def write_json(self, data: Dict[str, Any], name=None) -> Path:
```
Comment (BonusPlay) [inline teste2e/test_format_upgrades.py:27]:
```suggestion
    def write_json(self, data: Dict[str, Any], name: Optional[str] = None) -> Path:
```

This requires to `from typing import Optional`!

File: teste2e/test_format_upgrades.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+import json
+from typing import Dict, Any
+
+
+class TemporaryStorage:
+    def __init__(self):
+        self.tmpfiles = []
+        self.tmpdir = tempfile.gettempdir()
+
+    def tmpfile(self, name=None) -> Path:
```
Comment (BonusPlay) [inline teste2e/test_format_upgrades.py:19]:
```suggestion
    def tmpfile(self, name: Optional[str] = None) -> Path:
```

This requires to `from typing import Optional`!

File: teste2e/test_format_upgrades.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+            filepath = self.tmpdir + "/" + name
+        self.tmpfiles.append(filepath)
+        return Path(filepath)
+
+    def write_json(self, data: Dict[str, Any], name=None) -> Path:
+        p = self.tmpfile(name=name)
+        p.write_text(json.dumps(data))
+        return p
+
+    def write_text(self, text: str, name=None) -> Path:
```
Comment (BonusPlay) [inline teste2e/test_format_upgrades.py:32]:
```suggestion
    def write_text(self, text: str, name: Optional[str] = None) -> Path:
```

This requires to `from typing import Optional`!

**Pull request merged**